### PR TITLE
WIP: HDF5 'Create' Code Generation

### PR DESCRIPTION
**Pull request by hodger**
@scopatz 

PR for collaboration, still bugs to be worked out. This PR focused mainly on generating the HDF5Back::CreateTable function.

### CODE REVIEW

File: cli/cycpp.py [hunk 1]
```diff
@@ -492,6 +492,8 @@ def transform_pass2(self, statement, sep):
         annotations['type'] = state.canonize_type(vtype, vname,
                                                   statement=statement)
         annotations['index'] = len(state.context[classname]['vars'])
+        #added this
```
Comment (scopatz) [inline cli/cycpp.py]: PEP single space after `#`

File: cli/cycpp.py [hunk 2]
```diff
@@ -813,12 +815,58 @@ def canonize_type(self, t, name="<member variable>", statement=None):
...
+        elif len(current_shape) == expected_shape_length:
+            new_shape = current_shape
+        #Shape is too short.
+        elif len(current_shape) < expected_shape_length:    
+            diff = expected_shape_length - len(current_shape)
+            new_shape = current_shape.extend([-1] * diff)
+        #Shape is too long- we throw an error.
+        elif len(current_shape) > expected_shape_length:
+            err_string = ("Shape array for type {t} is not formatted correctly."
+                          + " Expected length {length} or less.")
```
Comment (scopatz) [inline cli/cycpp.py]: No need for `+` here since string will auto concatenate

File: src/hdf5_back_gen.py [hunk 3]
```diff
@@ -1075,11 +1074,459 @@ def prefixed_lines():
...
+    while i < len(result):
+        if isinstance(result[i], str):
+            i += 1
+        else:
+            temp = result[i][1:]
+            i += 1
+            for j in range(0, len(temp)):
+                result.insert(i+j, temp[j])
+    if isinstance(canon, str):
+        print(result)
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: Is this needed?

File: src/hdf5_back_gen.py [hunk 4]
```diff
@@ -1075,11 +1074,459 @@ def prefixed_lines():
...
+                                          args=[Raw(code="field_types[i]")])))
+    return body
+
+HDF5_PRIMITIVES = {"INT": "H5T_NATIVE_INT",
+              "DOUBLE": "H5T_NATIVE_DOUBLE",
+              "FLOAT": "H5T_NATIVE_FLOAT",
+              "BOOL": "H5T_NATIVE_CHAR",
+              "STRING": "CreateFLStrType({size})",
+              "BLOB": "sha1_type_",
+              "UUID": "uuid_type_"}
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: Odd alignment

File: src/hdf5_back_gen.py [hunk 5]
```diff
@@ -1088,7 +1535,95 @@ def main():
...
         output += CPPGEN.visit(case_template(type_node, read_x))
     output = indent(output, INDENT * 5)
-    print(output)
+    return output
+
+NOT_VL = []
+VARIATION_DICT = {}
+ORIGIN_DICT = {}
+
+io_error = Raw(code="throw IOError(\"the type for column \'\"+std::string(field_names[i])+\"\' is not yet supported in HDF5.\");")
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: Long line

File: src/hdf5_back_gen.py [hunk 6]
```diff
@@ -1088,7 +1535,95 @@ def main():
...
+
+raw_string = Raw(code="dbtypes[i]=STRING;\n"
+                      +"field_types[i]=H5Tcopy(H5T_C_S1);\n"
+                      +"H5Tset_size(field_types[i], shape[0]);\n"
+                      +"H5Tset_strpad(field_types[i], H5T_STR_NULLPAD);\n"
+                      +"opened_types_.insert(field_types[i]);\n"
+                      +"dst_sizes[i]=sizeof(char)*shape[0];\n")
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: `+` not needed, and also below

**Comment by scopatz** [comment]
Thanks @hodger! This is looking great to me.  Just needs the bug fixes :)

**Comment by hodger** [comment]
@scopatz I believe I took care of the spots you mentioned. The change I made to tools.py was because the nosetests failed to find lib.pyx. [This solution](http://stackoverflow.com/questions/7508803/how-do-i-import-function-from-pyx-file-in-python) was where I got the idea.

However, I'm still getting some nose failures on my machine. I think these are related to the new Python bindings, and since I'm not seeing the same failures on the CI checks, it's probably related to an out-of-date dependency. The error message is mainly `Operation only allowed in c++` for a bunch of the `del` and `new` calls in lib.pyx.

**Comment by scopatz** [comment]
Hi @hodger - cool, and it looks like this passes CI.  Try updating Cython, since that is where the error is coming from.

File: src/hdf5_back_gen.py [hunk 7]
```diff
@@ -1075,11 +1076,516 @@ def prefixed_lines():
...
+                    child_array = [child_array]
+                size += get_item_size(CANON_TO_NODE[t.canon[1]], child_array,
+                                      vl_flag=vl_flag, depth=depth+1)
+            size += ")"
+            if t.canon[0] in variable_length_types:
+                size += "*" + "shape[" + str(shape_array[0]) + "]"
+        size += ")"
+        return size    
+
+def H5Tarray_create2(item_variable, rank=1, dims="&shape0"):
```
Comment (scopatz) [inline src/hdf5_back_gen.py]: PEP8: function names are `lower_with_underscores`, same below

File: tests/tools.py [hunk 8]
```diff
@@ -15,6 +15,9 @@
 from nose.plugins.attrib import attr
 from nose.plugins.skip import SkipTest
 
+import pyximport
+pyximport.install()
```
Comment (scopatz) [inline tests/tools.py]: This should not be needed.

**Comment by hodger** [comment]
@scopatz fixed those spots. Found a solution to my nosetest problem, looks like I wasn't specifying the test directory correctly.

**Comment by scopatz** [comment]
Cool.

**Comment by scopatz** [comment]
OK this looks good to me @hodger. Let me know if you think it  is read y to merge, I hesitate because of the WIP in the title.

**Comment by hodger** [comment]
@scopatz I think this is ready! I'll end up opening a new PR fairly soon for the rest of the generation, so as long as that's OK with you, I'm all for merging this in.

**Comment by scopatz** [comment]
Sounds great, thanks!

**Pull request merged**